crates/ide-diagnostics/src/handlers/unlinked_file.rs

```diff
@@ -28,39 +28,40 @@ pub(crate) fn unlinked_file(
     file_id: FileId,
 ) {
     let mut range = TextRange::up_to(ctx.sema.db.line_index(file_id).len());
-    let fixes = fixes(ctx, file_id, range);
-    // FIXME: This is a hack for the vscode extension to notice whether there is an autofix or not before having to resolve diagnostics.
-    // This is to prevent project linking popups from appearing when there is an autofix. https://github.com/rust-lang/rust-analyzer/issues/14523
-    let message = if fixes.is_none() {
-        "This file is not included in any crates, so rust-analyzer can't offer IDE services."
+
+    let check_crate = ctx.sema.first_crate(file_id);
+    let is_in_crate = check_crate.is_some();
+
+    let is_crate_root = check_crate
+        .as_ref()
+        .map(|krate| krate.root_module().definition_source(ctx.sema.db).file_id == file_id)
+        .unwrap_or(false);
+
+    let is_unlinked = !is_in_crate || is_crate_root;
```

Comment:
I am confused by what you are doing here. If we end up with this diagnostic we have already figured out that the file is in fact unlinked so I am not sure what all of this is actually trying to compute here